imxrt-hal/src/tempmon.rs

```diff
@@ -0,0 +1,395 @@
+    /// hot_temp * 1000
+    hot_temp: i32,
+
+    /// measurement active flag to get measure_temp unblocking
+    measurement_active: bool,
+}
+
+impl TempMon {
+    /// converts the temp_cnt into a human readable temperature [°mC] (1/1000 °C)
+    fn convert(&self, temp_cnt: i32) -> i32 {
+        let n_meas = temp_cnt - self.hot_count;
+        self.hot_temp - n_meas * self.scaler
+    }
+
+    /// decode the temp_value into measurable bytes
+    ///
+    /// param **temp_value_mc**: in °mC (1/1000°C)
+    ///
+    fn decode(&self, temp_value_mc: i32) -> u32 {
+        let v = (temp_value_mc - self.hot_temp) / self.scaler;
+        (self.hot_count - v) as u32
+    }
+
+    /// triggers a new measurement and waits until it's finished
+    ///
+    /// If you configured automatically repeating, this will trigger additional measurement.
+    /// Use get_temp instate to get the last read value
+    ///
+    /// The returning temperature in 1/1000 Celsius (°mC)
+    ///
+    /// Example: 25500°mC -> 25.5°C
+    pub fn measure_temp(&mut self) -> nb::Result<i32, TempMonError> {
+        if !self.is_powered_up() {
+            Err(nb::Error::from(TempMonError::PowerDown))
+        } else {
+            // if no measurement is active, trigger new measurement
+            if !self.measurement_active {
+                ral::write_reg!(ral::tempmon, self.base, TEMPSENSE0, MEASURE_TEMP: 1);
+                self.measurement_active = true;
+            }
+
+            // if the measurement is not finished or not started
+            // i.MX Docs: This bit should be cleared by the sensor after the start of each measurement
+            if ral::read_reg!(ral::tempmon, self.base, TEMPSENSE0, FINISHED == 0) {
+                // this could be triggered again without any effect
+                Err(nb::Error::WouldBlock)
+            } else {
+                self.measurement_active = false;
+                let temp_cnt = ral::read_reg!(ral::tempmon, self.base, TEMPSENSE0, TEMP_CNT) as i32;
+                Ok(self.convert(temp_cnt))
+            }
+        }
+    }
+
+    /// Returns the last read value from the temperature sensor
+    ///
+    /// The returning temperature in 1/1000 Celsius (°mC)
+    ///
+    /// Example: 25500°mC -> 25.5°C
+    pub fn get_temp(&self) -> nb::Result<i32, TempMonError> {
+        if self.is_powered_up() {
+            let temp_cnt = ral::read_reg!(ral::tempmon, self.base, TEMPSENSE0, TEMP_CNT) as i32;
+            Ok(self.convert(temp_cnt))
+        } else {
+            Err(nb::Error::from(TempMonError::PowerDown))
+        }
+    }
+
+    /// Starts the measurement process. If the measurement frequency is zero, this
+    /// results in a single conversion.
+    pub fn start(&mut self) -> nb::Result<(), TempMonError> {
+        if self.is_powered_up() {
+            ral::write_reg!(
+                ral::tempmon,
+                self.base,
+                TEMPSENSE0,
+                MEASURE_TEMP: 1
+            );
```

Comment:
This `write_reg!` sets the `MEASURE_TEMP` bit high, _and sets all other register bits to zero_. We're similarly using `write_reg!` throughout this implementation to change one field in a register. Is this causing any issues in practice?

Would `modify_reg`, which preserves the unspecified fields, be more appropriate? Or, could we use the various `*_SET`, `*_CLR`, and `*_TOGGLE` registers with `write_reg!` to achieve our goals? It might be better to write this 1 to `TEMPSENSE0_SET`, since the hardware will preserve the other bits. (I'd recommend using the set/clear/toggle registers when possible. We could then discuss `&mut self` vs `&self` receivers for these methods.)

Reply:
I changed it, please let me know if this is how it should be. I didn't find an example.

I think it should be `&mut self.` 
Otherwise, somebody could change the behavior of the tempmon without having a mutable reference.

Reply:
> I changed it, please let me know if this is how it should be. I didn't find an example

Found it, `modify_reg!()` is what I was looking for

Reply:
I just changed it, because I didn't really understand how it works. but using `*_SET` and `*_CLR` feels much faster!